text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,267 @@
+---
+feature name: CDK Metadata v2
+start date: 2020-09-23
+rfc pr: (leave this empty)
+related issue: #253
+---
+
+# Abstract
+
+The CDK Metadata resource will start tracking individual constructs used in a
+Stack's synthesis. For every construct used in a Stack from an AWS-authored
+library we will collect the library name, version and qualified class name
+(using a generated lookup table to obtain submodule names) and submit these
+to the metadata resource in a gzipped blob.
+
+# Motivation
+
+The goal of CDK Metadata is twofold:
+
+* Track what library versions people are using, so that if there are severy (security)
+  problems with a library, we can contact them in a targeted fashion and suggest
+  they upgrade.
+* See what construct libraries are being used by people to inform future development
+  activities (example insight: are there areas where people are using L1s
+  instead of L2s, signaling the need to invest more effort into L2
+  development).
+
+The current CDK Metadata is based on libraries loaded into the NodeJS process. When
+we move to CDK v2, the entire CDK library will be vended as one library, destroying
+our ability to measure in the same way.
+
+At the same time, our current method fails in the face of multiple distinct
+stacks in the same application. For example, in the case of a CDK Pipelines
+application, the `@aws-cdk/pipelines` library would be counted `N+1` times:
+in addition to the pipeline stack itself, once for every environment deployed
+via that pipeline because the library happened to be in memory while those
+stacks were being synthesized.
+
+We therefore need to move to a more fine-grained tracking system: a stack
+will report the individual constructs used to synthesize that stack.
+
+We will retain the existing filter that we will only collect information
+about constructs from AWS-authored construct libraries. 3rd-party information
+will not be collected.
+
+# Design
+
+There are two subproblems to address:
+
+* Collecting construct identifiers: library name, version, and qualified class name.
+* Encoding that information into the metadata resource.
+
+## Collecting construct identifiers
+
+We need a sufficiently unique string representation of a construct('s class). Since
+we don't have manual annotation of construct class types at the moment, we will
```

Comment:
I don't mean anything. I literally only mean that a human must type some code somewhere to indicate the construct type. Haven't gotten into the mechanism yet and since I don't think we want any solution that requires human annotation, I don't want to spend any more cycles on thinking up a plausible mechanism to do this only to then discard it.